substrate/client/network/src/litep2p/ipfs_dht.rs

```diff
@@ -0,0 +1,218 @@
+
+		(Self { kademlia_handle, block_provider }, config)
+	}
+
+	pub async fn run(mut self) {
+		let mut changes = self.block_provider.changes();
+		let mut inflight_queries = HashMap::new();
+		let mut random_walk_query = None;
+		let mut random_walk_interval = tokio::time::interval(RANDOM_WALK_INTERVAL);
+		random_walk_interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+		loop {
+			tokio::select! {
+				change = changes.next(), if inflight_queries.len() < MAX_INFLIGHT_QUERIES => {
+					match change {
+						None => {
+							debug!(target: LOG_TARGET, "BlockProvider terminated, terminating IpfsDht");
+							return
+						},
+						Some(Change::Added(multihash)) => {
+							let key = RecordKey::new(&multihash.to_bytes());
+
+							trace!(
+								target: LOG_TARGET,
+								"IPFS DHT start providing key: {}, CID: {}",
+								HexDisplay::from(&key.as_ref()),
+								Cid::new_v1(RAW_CODEC, multihash),
+							);
+
+							let query_id = self.kademlia_handle.start_providing(key, QUORUM).await;
+							inflight_queries.insert(query_id, multihash);
+						},
+						Some(Change::Removed(multihash)) => {
+							let key = RecordKey::new(&multihash.to_bytes());
+
+							trace!(
+								target: LOG_TARGET,
+								"IPFS DHT stop providing key: {}, CID: {}",
+								HexDisplay::from(&key.as_ref()),
+								Cid::new_v1(RAW_CODEC, multihash),
+							);
+
+							self.kademlia_handle.stop_providing(key).await;
+						},
+					}
+				},
+				_ = random_walk_interval.tick() => {
+					if random_walk_query.is_some() {
+						// Do not start a new random walk if the previous one hasn't finished.
+						continue;
+					}
+
+					random_walk_query = Some(self.kademlia_handle.find_node(PeerId::random()).await);
+				},
+				event = self.kademlia_handle.next() => {
+					match event {
+						None => {
+							debug!(target: LOG_TARGET, "IPFS Kademlia terminated, terminating IpfsDht");
+							return
+						}
+						Some(KademliaEvent::AddProviderSuccess { query_id, provided_key }) => {
+							if let Some(multihash) = inflight_queries.remove(&query_id) {
+								trace!(
+									target: LOG_TARGET,
+									"IPFS DHT provider publish success, key: {}, CID: {}",
+									HexDisplay::from(&provided_key.as_ref()),
+									Cid::new_v1(RAW_CODEC, multihash),
+								);
+							} else {
+								trace!(
+									target: LOG_TARGET,
+									"IPFS DHT provider refresh success, key: {}",
+									HexDisplay::from(&provided_key.as_ref()),
+								);
+							}
+						},
+						Some(KademliaEvent::FindNodeSuccess { query_id, peers, .. }) => {
+							debug_assert_eq!(Some(query_id), random_walk_query);
+							trace!(target: LOG_TARGET, "DHT random walk yielded {} peers", peers.len());
+
+							random_walk_query = None;
+						},
+						Some(KademliaEvent::QueryFailed { query_id }) => {
+							if Some(query_id) == random_walk_query {
+								trace!(target: LOG_TARGET, "DHT random walk failed");
+
+								random_walk_query = None;
```

Comment:
For the following example, we'll wait quite a bit without issuing a new query:
- T0: query submited
- T1 (10seconds later): query fails
- T2: (9 minutes 50 seconds later): we resubmit the query

Could we initiate the random walk on this failure instead? This way we always have an inflight query until we can find peers